tools/tm-signer-harness/internal/test_harness_test.go

```diff
@@ -0,0 +1,207 @@
+		"value": "ZCsuTjaczEyon70nmKxwvwu+jqrbq5OH3yQjcK0SFxc="
+	},
+	"priv_key": {
+		"type": "tendermint/PrivKeyEd25519",
+		"value": "8O39AkQsoe1sBQwud/Kdul8lg8K9SFsql9aZvwXQSt1kKy5ONpzMTKifvSeYrHC/C76Oqturk4ffJCNwrRIXFw=="
+	}
+}`
+
+	stateFileContents = `{
+	"height": "0",
+	"round": "0",
+	"step": 0
+}`
+
+	genesisFileContents = `{
+	"genesis_time": "2019-01-15T11:56:34.8963Z",
+	"chain_id": "test-chain-0XwP5E",
+	"consensus_params": {
+		"block_size": {
+		"max_bytes": "22020096",
+		"max_gas": "-1"
+		},
+		"evidence": {
+		"max_age": "100000"
+		},
+		"validator": {
+		"pub_key_types": [
+			"ed25519"
+		]
+		}
+	},
+	"validators": [
+		{
+		"address": "D08FCA3BA74CF17CBFC15E64F9505302BB0E2748",
+		"pub_key": {
+			"type": "tendermint/PubKeyEd25519",
+			"value": "ZCsuTjaczEyon70nmKxwvwu+jqrbq5OH3yQjcK0SFxc="
+		},
+		"power": "10",
+		"name": ""
+		}
+	],
+	"app_hash": ""
+}`
+
+	defaultAcceptDeadline = 100
+	defaultConnDeadline   = 100
+)
+
+func TestRemoteSignerTestHarnessMaxAcceptRetriesReached(t *testing.T) {
+	cfg := makeConfig(t, 1, 2)
+	defer cleanup(cfg)
+
+	th, err := NewTestHarness(log.TestingLogger(), cfg)
+	assert.NoError(t, err)
+	donec := make(chan struct{})
+	go func() {
+		defer close(donec)
+		th.Run()
+	}()
+	<-donec
+	assert.Equal(t, ErrMaxAcceptRetriesReached, th.exitCode)
+}
+
+func TestRemoteSignerTestHarnessSuccessfulRun(t *testing.T) {
+	harnessTest(
+		t,
+		func(th *TestHarness) *privval.RemoteSigner {
+			return newMockRemoteSigner(t, th, th.fpv.Key.PrivKey, false, false)
+		},
+		NoError,
+	)
+}
+
+func TestRemoteSignerPublicKeyCheckFailed(t *testing.T) {
+	harnessTest(
+		t,
+		func(th *TestHarness) *privval.RemoteSigner {
+			return newMockRemoteSigner(t, th, ed25519.GenPrivKey(), false, false)
+		},
+		ErrTestPublicKeyFailed,
+	)
+}
+
+func TestRemoteSignerProposalSigningFailed(t *testing.T) {
+	harnessTest(
+		t,
+		func(th *TestHarness) *privval.RemoteSigner {
+			return newMockRemoteSigner(t, th, th.fpv.Key.PrivKey, true, false)
+		},
+		ErrTestSignProposalFailed,
+	)
+}
+
+func TestRemoteSignerVoteSigningFailed(t *testing.T) {
+	harnessTest(
+		t,
+		func(th *TestHarness) *privval.RemoteSigner {
+			return newMockRemoteSigner(t, th, th.fpv.Key.PrivKey, false, true)
+		},
+		ErrTestSignVoteFailed,
+	)
+}
+
+func newMockRemoteSigner(t *testing.T, th *TestHarness, privKey crypto.PrivKey, breakProposalSigning bool, breakVoteSigning bool) *privval.RemoteSigner {
+	return privval.NewRemoteSigner(
+		th.logger,
+		th.chainID,
+		types.NewMockPVWithParams(privKey, breakProposalSigning, breakVoteSigning),
+		privval.DialTCPFn(
+			th.addr,
+			time.Duration(defaultConnDeadline)*time.Millisecond,
+			ed25519.GenPrivKey(),
+		),
+	)
+}
+
+// For running relatively standard tests.
+func harnessTest(t *testing.T, rsMaker func(th *TestHarness) *privval.RemoteSigner, expectedExitCode int) {
+	cfg := makeConfig(t, 100, 3)
+	defer cleanup(cfg)
+
+	th, err := NewTestHarness(log.TestingLogger(), cfg)
+	assert.NoError(t, err)
```

Comment:
better to use require here to stop the test (because next steps will probably fail to do -> make no sense to continue)